### PR TITLE
Allows blacklisting words from IC chat via config

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

~~Blocks netspeak in IC using regex.~~ Implements a config file for filtering IC, for the purpose of blocking netspeak. It's still allowed in asay/dsay/ooc/etc. I'm aware that this list is not exhaustive and that it can be fairly easily bypassed. But this blocks most of the common occurrences so it'll make life easier for admins, and it'll be obvious when people are intentionally breaking the rule by bypassing it.

It also doesn't block words that contain blacklisted words in them. For example, "lol" (if configured) would be blocked but "lollygagging" would not.

I expect this to cause at least three (3) angry rants on reddit.

## Why It's Good For The Game

Netspeak in IC is against the rules. MUH IMMERSHUN, and all that.

## Changelog
:cl: ike709
add: IC chat can now be filtered via config to prevent things like netspeak or anything else against the rules.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/say.dm
Comment (Wilchenx) [inline code/modules/mob/living/say.dm]: add "reddit"

**Comment by Fikou** [comment]
what about gibbering messages

File: code/modules/mob/living/say.dm
Comment (ike709) [reply code/modules/mob/living/say.dm]: Don't tempt me.

**Comment by Wilchenx** [comment]
A valid concern is clown name policy, if the player was to use one of the blacklisted words as their name, noone would be able to call them out on the radio.

**Review by redmoogle — CHANGES_REQUESTED** [review]
words to also include
'netspeak'
'ooc'
'ic'
'ROF(T)L'
'Rly'
'pls'
'ez'
'cuz'
not all but should help alot

**Comment by nicbn** [comment]
Well it's not that rare to see (hear) someone saying something like lol irl, that's really dumb

**Comment by redmoogle** [comment]
why not ban the world netspeak itself

> Well it's not that rare to see (hear) someone saying something like lol irl, that's really dumb

yes but it's frowned upon in-game because you should say haha instead or something like that

File: code/modules/mob/living/say.dm
Comment (Tlaltecuhtli) [reply code/modules/mob/living/say.dm]: add "owo" and "uwu" as they are one of the major crimes of this kind

**Comment by FloranOtten** [comment]
lol and lmao are words used IRL  as well. You may call it cringe but there you go. I think I hear the word lmao spoken more than I hear the entire sentence spoken.

**Comment by HomocidlePancakes** [comment]
>lol and lmao are words used IRL as well. You may call it cringe but there you go. I think I hear the word lmao spoken more than I hear the entire sentence spoken.

Sorry. You have to be atleast 18 to play on TG. How did you end up in this unwholesome hellhole child?

**Comment by partykp** [comment]
What if you made all the instances of "laughing" words like lol and lmao just make the offender say *laugh

**Comment by nervere** [comment]
Is it possible for you to make what words are filtered a config option?
What about a config option with the ability to use regex of our own?
Either works, I just feel this being hardcoded is not a good idea.

**Comment by OverlyGeneric** [comment]
I don't believe this will cause much trouble on Reddit. they may actually commend you for relieving their frustrations.

good work.

**Comment by nervere** [comment]
> I don't believe this will cause much trouble on Reddit. they may actually commend you for relieving their frustrations.
> 
> good work.

No one cares about Reddit.

**Comment by JJRcop** [comment]
This should be a config option in the interest of downstreams

File: code/modules/mob/living/say.dm
Comment (optimumtact) [inline code/modules/mob/living/say.dm]: does this actually ensure the regex is only compiled once?

Also I suspect a chain of findtext calls will actually be faster.

third this should be configurable list in config

@MrStonedOne are you happy for a wordfilter to be applied?

File: code/modules/mob/living/say.dm
Comment (optimumtact) [reply code/modules/mob/living/say.dm]: please don't use code comments unless you are pointing out an actual code issue or are a maintainer please.

They're otherwise wildly disruptive to the rest of the conversation.

File: code/modules/mob/living/say.dm
Comment (4dplanner) [inline code/modules/mob/living/say.dm]: should allow you to force it with a warning imo

Comment (optimumtact) [reply code/modules/mob/living/say.dm]: why would we willingly let people say something that can net them a 5 minute ban?

Comment (4dplanner) [reply code/modules/mob/living/say.dm]: There are legitimate cases, like saying "my nephew posted `lol i rekt a scrub` on facebook!", in which you're referring not using

**Comment by Wilchenx** [comment]
>this being hardcoded is not a good idea.

Absolutely correct, downstreams may not want a fixed list.

File: code/modules/mob/living/say.dm
Comment (MrStonedOne) [reply code/modules/mob/living/say.dm]: Make it a configurable list and kick the bucket down to the headmins, but this would need to be changed as static vars are inited with global vars before anything else happens, including config loading.

Comment (ike709) [reply code/modules/mob/living/say.dm]: I don't know how to check if the regex is only compiled once.

But I'll profile this vs findtext and rework it into a config when I have time, if you are willing to merge this once those improvements have been made. Otherwise I won't waste my time.

File: code/modules/mob/living/say.dm
Comment (optimumtact) [reply code/modules/mob/living/say.dm]: also stop using code comments on things that aren't not directly technical feedback

**Comment by ike709** [comment]
I'll rework this into a config later today when I have free time.

**Comment by bgobandit** [comment]
you are absolutely certain that this blocks only the words themselves and not words containing those character strings, correct?

**Comment by MrStonedOne** [comment]
The `\\b` matches a space

File: code/modules/mob/living/say.dm
Comment (optimumtact) [reply code/modules/mob/living/say.dm]: Im not opposed to it

make sure you add a config option to globally enable/disable it as well, or just use the empty state of the list to skip any checks

File: code/modules/mob/living/say.dm
Comment (JJRcop) [inline code/modules/mob/living/say.dm]:
```suggestion
	if(!forced && findtext(message, netspeak))
```

File: code/modules/mob/living/say.dm
Comment (ike709) [reply code/modules/mob/living/say.dm]: Turned this into a config.

**Comment by ike709** [comment]
It's a config file now.

File: code/modules/mob/living/say.dm
Comment (JJRcop) [inline code/modules/mob/living/say.dm]:
```suggestion
		if(!forced && findtext(message, filter))
```

**Comment by ike709** [comment]
Unless any other requested changes arise, I think this is ready.

File: code/modules/mob/living/say.dm
Comment (ike709) [reply code/modules/mob/living/say.dm]: Done.

File: code/modules/mob/living/say.dm
Comment (optimumtact) [inline code/modules/mob/living/say.dm]: you need to compile this once at world init please

Comment (ike709) [reply code/modules/mob/living/say.dm]: Done.

File: code/modules/mob/living/say.dm
Comment (nervere) [inline code/modules/mob/living/say.dm]: this should probably be a configurable message as well

**Review by nervere — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/say.dm
Comment (ike709) [reply code/modules/mob/living/say.dm]: People can already click the configurable rules button at the top of the game window, but I will alter the message to mention it.

**Comment by YakumoChen** [comment]
isn't this more of an administrative issue? why do we need to hardcode this

**Comment by FloranOtten** [comment]
I am also not a fan of having hardcoded filters. We don't need god damn ban bots. Who exactly was so bothered by someone saying 'lol'? Fuck, I've heard people say that in common speech.

**Comment by Alek2ander** [comment]
The filter shouldn't apply to the binary channel because it's literally netspeak

**Comment by optimumtact** [comment]
@YakumoChen apparently you're so stuck in your bunker you haven't seen the drama going on right now

**Comment by Mark9013100** [comment]
lol

**Comment by dr-bee** [comment]
Wouldnt it be a nicer solution to have LOL just trigger the laughing emote and not the word lol instead of blocking the entire message?

**Comment by FloranOtten** [comment]
> @YakumoChen apparently you're so stuck in your bunker you haven't seen the drama going on right now

Drama?

**Comment by Tlaltecuhtli** [comment]
it might break immersion more as for stuff like force speech nanites on "lol" and having everyone spammed by unknown ooc in ic warns

this change is dumb let the banbots handle their dopamine level by giving out 5 min bans

**Comment by CPTANT** [comment]
wtf and omg are just the written abbreviation of what the fuck and oh my God . People say that all the time. 

And lol is used all the time here.

**Comment by ike709** [comment]
Imagine unironically saying "lol" as an adult.

**Comment by CPTANT** [comment]
Well "lol" has been the Dutch word for fun since a couple of centuries it's not that weird.

**Comment by Wilchenx** [comment]
We're a english speaking community

**Comment by Yttriums** [comment]
Is netspeak more immersion breaking than having players get popups when they say specific disallowed words?

**Comment by Wilchenx** [comment]
Lifeweb weekbans you for using netspeak

So i'd say yes, its frowned upon to use netspeak while roleplaying.

**Comment by CPTANT** [comment]
We're a space station 13 community

**Comment by Wilchenx** [comment]
Ale jakie ma to znaczenie?

**Comment by 4dplanner** [comment]
> it might break immersion more as for stuff like force speech nanites on "lol" and having everyone spammed by unknown ooc in ic warns
> 
> this change is dumb let the banbots handle their dopamine level by giving out 5 min bans

that's why it checks the forced variable

**Comment by JJRcop** [comment]
The title of the PR should be changed since recent additions to it have modified what it does, and the title isn't accurate anymore.

After recent changes, the PR adds the ability for server administrators to blacklist words, but does not come preloaded with any words anymore.

**Comment by chaoko99** [comment]
"Cuz" is something that isn't netspeak... It's short for "because", and real people have been saying it for ages.

**Comment by Yttriums** [comment]
The game takes place in a future with advanced electronic communication, obviously electronic communication will have influenced the way people speak just like it did in our world.

Netspeak IC is way more plausible than, say, having a clown on a corporate spaceship for entertaining adults in 2500ad.

**Comment by Wilchenx** [comment]
surely noone will speak about reddit in 2500ad

**Comment by ike709** [comment]
@JJRcop Good point.

**Comment by ike709** [comment]
@CPTANT Well then all of the Dutch servers can just not put "lol" in the config file.

**Comment by ike709** [comment]
@chaoko99 Then I guess it's good that "cuz" was never blacklisted at any point even when it was hardcoded.

**Comment by nervere** [comment]
If you cannot talk without using netspeak, I'm afraid this roleplaying game is not for you.

**Comment by optimumtact** [comment]
this is a configurable thing if you want to argue about netspeak on tg please go to the forum policy discussion

**Pull request merged**
